### PR TITLE
Feat/sym log ticks

### PR DESCRIPTION
**Pull request by skie1997**
[[中文版模板 / Chinese template](https://github.com/VisActor/Vutil/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Chore
- [ ] Other (about what?)

### 🔗 Related issue link
visactor/vchart#508

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vscale/src/log-scale.ts [hunk 1]
```diff
@@ -47,9 +49,8 @@ export class LogScale extends ContinuousScale {
 
     const logs = logp(this._base);
     const pows = powp(this._base);
-    const domain = this._niceDomain ?? this._domain;
 
-    if (domain[0] < 0) {
+    if (this._domain[0] < 0) {
```
Comment (xile611) [inline packages/vscale/src/log-scale.ts:53]: nice相关处理需要保留一下，因为nice会影响domain

File: packages/vscale/src/log-scale.ts [hunk 2]
```diff
@@ -156,48 +159,16 @@ export class LogScale extends ContinuousScale {
...
-        return this;
-      }
-    } else {
-      niceType = 'all';
-    }
-
-    if (niceType) {
-      const niceDomain = nice(originalDomain.slice(), {
+  nice(): this {
+    return this.domain(
```
Comment (xile611) [inline packages/vscale/src/log-scale.ts:166]: nice相关处理需要保留

File: packages/vscale/src/log-scale.ts [hunk 3]
```diff
@@ -147,7 +150,7 @@ export class LogScale extends ContinuousScale {
    */
   forceTicks(count: number = 10): any[] {
     const d = this.calculateVisibleDomain(this._range);
-    return forceTicks(d[0], d[d.length - 1], count);
+    return forceTicksBaseTransform(d[0], d[1], count, this.transformer, this.untransformer);
```
Comment (xile611) [inline packages/vscale/src/log-scale.ts:153]: don't read `d[1]`, just use `d[d.length - 1]`, in same case ,domain has length > 2

File: packages/vscale/src/log-scale.ts [hunk 4]
```diff
@@ -156,48 +159,16 @@ export class LogScale extends ContinuousScale {
    */
   stepTicks(step: number): any[] {
     const d = this.calculateVisibleDomain(this._range);
-    return stepTicks(d[0], d[d.length - 1], step);
+    return forceTicksBaseTransform(d[0], d[1], step, this.transformer, this.untransformer);
```
Comment (xile611) [inline packages/vscale/src/log-scale.ts:162]: don't read `d[1]`, just use `d[d.length - 1]`, in same case ,domain has length > 2

File: packages/vscale/package.json [hunk 5]
```diff
@@ -29,7 +29,7 @@
     "build": "bundle --clean",
     "dev": "bundle --clean -f es -w",
     "start": "vite ./vite",
-    "test": "jest",
+    "test": "jest __tests__/symLog.test.ts",
```
Comment (xile611) [inline packages/vscale/package.json:32]: remove this line

File: packages/vscale/src/symlog-scale.ts [hunk 6]
```diff
@@ -36,4 +38,71 @@ export class SymlogScale extends LinearScale {
...
+  /**
+   * 基于给定step的ticks数组生成
+   * @param step
+   */
+  stepTicks(step: number): any[] {
+    const d = this.calculateVisibleDomain(this._range);
+    return forceTicksBaseTransform(d[0], d[1], step, this.transformer, this.untransformer);
+  }
+
+  nice(): this {
```
Comment (xile611) [inline packages/vscale/src/symlog-scale.ts:66]: keep the same as logScale

File: packages/vscale/src/symlog-scale.ts [hunk 7]
```diff
@@ -36,4 +38,71 @@ export class SymlogScale extends LinearScale {
...
+    return ticksBaseTransform(d[0], d[1], count, this._const, this.transformer, this.untransformer);
+  }
+
+  /**
+   * 生成tick数组，这个tick数组的长度就是count的长度
+   * @param count
+   */
+  forceTicks(count: number = 10): any[] {
+    const d = this.calculateVisibleDomain(this._range);
+    return forceTicksBaseTransform(d[0], d[1], count, this.transformer, this.untransformer);
```
Comment (xile611) [inline packages/vscale/src/symlog-scale.ts:54]: don't read `d[1]`, `d[d.length - 1]` is better

File: packages/vscale/src/symlog-scale.ts [hunk 8]
```diff
@@ -36,4 +38,71 @@ export class SymlogScale extends LinearScale {
...
+  }
+
+  /**
+   * 基于给定step的ticks数组生成
+   * @param step
+   */
+  stepTicks(step: number): any[] {
+    const d = this.calculateVisibleDomain(this._range);
+    return forceTicksBaseTransform(d[0], d[1], step, this.transformer, this.untransformer);
```
Comment (xile611) [inline packages/vscale/src/symlog-scale.ts:63]: don't read `d[1]`, `d[d.length - 1]` is better

File: packages/vscale/src/utils/tick-sample.ts [hunk 9]
```diff
@@ -450,3 +471,65 @@ export function parseNiceOptions(originalDomain: number[], option: NiceOptions)
...
+    untransformer: TransformType
+  ) => {
+    const ticksResult: number[] = [];
+    const ticksMap = {};
+    const startExp = transformer(start);
+    const stopExp = transformer(stop);
+    let ticksExp = [];
+    // get ticks exp
+    if (Number.isInteger(base)) {
+      ticksExp = new LinearScale().domain([startExp, stopExp]).ticks(count);
```
Comment (xile611) [inline packages/vscale/src/utils/tick-sample.ts:491]: dont create `LinearScale` here, just call `ticks()` function

this line will produce circular dependencies

**Review by xile611 — CHANGES_REQUESTED** [review]

**Review by xuefei1313 — APPROVED** [review]

File: packages/vscale/src/log-scale.ts [hunk 10]
```diff
@@ -207,11 +211,12 @@ export class LogScale extends ContinuousScale {
   niceMin(): this {
     const maxD = this._domain[this._domain.length - 1];
     this.nice();
-    const niceDomain = this._domain.slice();
+    const niceDomain = cloneDeep(this._domain);
```
Comment (xile611) [inline packages/vscale/src/log-scale.ts:214]: `cloneDeep` may have worse performance than `slice()`

**Review by xile611 — APPROVED** [review]

**Pull request merged**